### PR TITLE
RFC: Asset Loader Service

### PR DESCRIPTION
**Pull request by trentmwillis**
Pre-reading/context: [Asset Manifest RFC](https://github.com/emberjs/rfcs/pull/153)

[Rendered](https://github.com/trentmwillis/rfcs/blob/asset-loader-service/text/0000-asset-loader-service.md).

### CODE REVIEW

**Comment by rwjblue** [comment]
Where are we at with this? 

@dgeb / @MiguelMadero / @nathanhammond / @trentmwillis - Have y'all reviewed?

**Comment by trentmwillis** [comment]
We have a working implementation in [ember-asset-loader](https://github.com/trentmwillis/ember-asset-loader/) that has been used successfully for both lazy loading Engines and generic use cases (e.g., loading a charting lib).

Primary open question to me is where does this Service live in the future? For many use cases, having a standalone, core-maintained addon seems sufficient. But, for the lazy Engines use case, it needs to integrate with some of the internals of Ember's Router, which makes it seem like this could potentially be part of Ember itself. At the very least, if it continues as a separate addon, we likely need a solid API for hooking into the Router.

**Comment by nathanhammond** [comment]
Have reviewed the RFC and many of the PRs that went into the implementation. This is set by my estimation.

**Comment by trentmwillis** [comment]
@rwjblue clarified some wording. I think this is ready to move to FCP.

**Comment by rwjblue** [comment]
We discussed this during todays Ember core team meeting today, and we are 👍 on this moving forward.

Given that there is little conversation here, and the core team is in favor I an moving this forward into the final comment period.

File: text/0000-asset-loader-service.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+};
+```
+
+### `loadAsset`
+
+The `loadAsset` method loads a single asset into the application. It accepts an `asset` object which contains `uri` and `type` fields, both of which are strings. The `type` determines which loading method will be called internally according to the type of asset being loaded and the `uri` will be passed as an argument to said loading method to specify the actual asset to load.
+
+`loadAsset` will return an `AssetPromise` that will resolve when the asset is successfully loaded or reject when the asset fails to load.
+
+Subsequent calls to `loadAsset` with the same arguments will return the same `AssetPromise` object to ensure a single source of truth for determining asset state as well as to prevent duplicate requests being made for the same asset.
```
Comment (courajs) [inline text/0000-asset-loader-service.md]: Can we clarify here that this is true unless/until `retryLoad` is called, and link down to the `Provide A Way To Attempt Recovery` section?

**Comment by rwjblue** [comment]
Discussed this a bit with @tomdale and we decided that we should close this. Not as "wontfix" but more as "already done" in addon space without needing more detailed core framework involvement. This is absolutely public API from ember-engines point of view, and it will be supported as such.

@trentmwillis - Thank you very much for your hard work on this, and I'm very sorry that we let this sit for _sooo long_.

**Pull request closed without merging**